crates/sqlexec/src/background_jobs.rs

```diff
@@ -0,0 +1,326 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use tokio::{
+    sync::{mpsc, oneshot},
+    task::JoinHandle,
+    time::Instant,
+};
+
+use crate::errors::{ExecError, Result};
+
+/// Collection of background jobs.
+pub struct JobRunner {
+    send: mpsc::UnboundedSender<ChannelMessage>,
+    listen: Arc<Mutex<Option<JoinHandle<()>>>>,
+}
+
+struct JobHandle {
+    join: JoinHandle<()>,
+    sleep: JoinHandle<()>,
+}
+
+enum ChannelMessage {
+    NewJob(BackgroundJob),
+    JobComplete(String, Result<()>),
+    HealthCheck,
+    Close,
+}
+
+impl JobRunner {
+    /// Create a new collection of background jobs.
+    pub fn new() -> Self {
+        let (send, recv) = mpsc::unbounded_channel();
+        // Spawn the listener for receiving messages.
+        let send_clone = send.clone();
+        let listen = tokio::spawn(async move { Self::listen(send_clone, recv).await });
+        let listen = Arc::new(Mutex::new(Some(listen)));
+        Self { send, listen }
+    }
+
+    /// Listens on the channel for new messages.
+    async fn listen(
+        send: mpsc::UnboundedSender<ChannelMessage>,
+        mut recv: mpsc::UnboundedReceiver<ChannelMessage>,
+    ) {
+        // Sends a "health check" mesage to the channel every two minutes.
+        let send_clone = send.clone();
+        tokio::spawn(async move {
+            let send = send_clone;
+            let mut health_check = tokio::time::interval(Duration::from_secs(2 * 60));
+            loop {
+                let _ = health_check.tick().await;
+                send.send(ChannelMessage::HealthCheck).unwrap();
+            }
+        });
+
+        let mut jobs: HashMap<String, JobHandle> = HashMap::new();
+        while let Some(msg) = recv.recv().await {
+            match msg {
+                ChannelMessage::NewJob(job) => {
+                    let job_name = job.name;
+
+                    if jobs.contains_key(&job_name) {
+                        continue;
+                    }
```

Comment:
Is it expected that we'll have unique job names per session? E.g. what would happen if session 1 submits a job for dropping a table, and session 2 submits a job for dropping a table?

Reply:
I imagined jobs would be named consistently.

So dropping a table `tab` from any session would create a job `"drop_table_tab"`. If both sessions drop the same table, they'll be batched.

And unique job names are not per session; they are per engine. Maybe each job has a method `prefix()` or simply `name()` (that's created from options) so they are consistent, and it's not up to the user to provide a name.

Reply:
Seems reasonable. Since everything pretty much has a UUID with it, we can just stick that in the job name, so no concerns here.